docs/src/pages/ThemingSpecification.mdx

```diff
@@ -0,0 +1,301 @@
+
+If new variants are added in the new theme, then the clients should be able to compose the existing component, by adding new design terms (via the props API). If the new theme doesn't need to support any new variant, nothing should be required from the client's side.
+
+##### Variants
+
+Variants are a way of allowing the users to specify new design terms for their components (among other customizations).
+
+In the CSS scenario, it is basically a way of allowing user to specify mapping to className for a new prop.
+
+In the CSSinJS scenario, it is allowing user to have new properties (design terms) inside the styles functions.
+
+#### Themes should cascade and merge allowing subtree themes to be updated at runtime and have the ability to adjust the incoming theme.
+
+If some parameters/tokens are changed on the root of the application, the changes should be propagated to all nested subtrees.
+
+#### RTL and disable animation modes should be configurable for the entire application or a subtree of the application.
+
+It is important to allow the users to specify the RTL/LTR mode for the entire application once, but if necessary they can specify it only in a subtree of the application (for example one ChatMessage)
+
+The same is true for the disable animations mode. It is important to allow the users to specify it for the entire application once, but if necessary they can specify it only in a subtree of the application (for example the `Loader` component should always be animated).
+
+#### Parent components should be able to customize the styles for the child components
+
+The parent component should be able to customize the styles for it's children/slots, usually for positioning scenarios, or contextual changes that should be applied when some design term is applied on the parent component. This is especially required for "recursive component"'s scenarios
+
+## Technical proposal
+
+In the rest of the specification, there will be some proposals of how the themes, mappings and utilities can be structured.
+
+#### CSS stylesheets usage
+
+Users should be able to define mapping from props to classNames per component level.
+
+This is inspired by the [React Semantic UI's className builders](https://github.com/Semantic-Org/Semantic-UI-React/blob/master/src/lib/classNameBuilders.js)
+
+Here is an example of how the mappings for a FluentButton component can look like:
+
+```
+const fluentButtonMappings = props => {
+  const { size, circular, attached, corner } = props
+  return cx(
+    size, // the value (small, medium, big)
+    useKeyOnly(circular, 'circular'),
+    useKeyOrValueAndKey(attached, 'attached'), // true or the value of the attached prop
+    useValueAndKey(corner, 'corner'), // corner="left" will create the classNames `corner left`
+  )
+}
+
+const FluentButton = compose(Button, {
+  displayName: 'FluentButton',
+  mapPropsToClasses: fluentButtonMappings,
+})
```

Comment:
Not sure that this a good example as it hides complexity, i.e. how slots should work? Let's remove it for now.

Reply:
I've expanded the example to contain slots too. We can iterate on the actual API later. Is there a strong reason why we should remove it?

Reply:
There is still one more confusing thing: `mapPropsToClasses: fluentButtonMappings,`.

- How it is different from `mapPropsToTerms`? I don't have design terms in CSS?
- How it is different from CSSinJS? In CSSinJS we are not coupling components with actual styles, but here we do it.

Reply:
I see what you're saying, maybe these should be somehow configured at a theme level. Will remove it for now.